common/retry_delays.go

```diff
@@ -0,0 +1,134 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/agglayer/aggkit/config/types"
+)
+
+var (
+	// If this error is in the returned error, it means that the function should be aborted
+	// and no more retries should be attempted.
+	ErrAbort          = fmt.Errorf("abort")
+	ErrInvalidConfig  = fmt.Errorf("invalid retry delays config")
+	ErrExecutionFails = fmt.Errorf("execution fails after retries exceeded")
+)
+
+// RetryDelays is a struct that holds the retry delays and the maximum number of retries.
+// It implements the RetryDelayer interface, which allows executing a function with retry logic.
+// The delays are specified as a slice of types.Duration, and the maximum number of retries can be set.
+// If maxRetries is set to 0, it means infinite retries are allowed.
+//
+// You can also abort the retrying wrapping the error ErrAbort into the result
+// this is useful if there are some conditions that should not be retried
+type RetryDelays struct {
+	Delays []types.Duration `mapstructure:"Delays"`
+	// MaxAttempts is the maximum number of retries to attempt.
+	// if MaxAttempts is 0, it means infinite retries.
+	// if MaxAttempts is 1, it means no retries will be attempted.
+	MaxAttempts int `mapstructure:"MaxAttempts"`
+}
+
+func (r *RetryDelays) Validate() error {
+	if r == nil {
+		return fmt.Errorf("retry delays cannot be nil. %w", ErrInvalidConfig)
+	}
+	if len(r.Delays) == 0 {
+		return fmt.Errorf("retry delays cannot be empty. %w", ErrInvalidConfig)
+	}
+	for _, delay := range r.Delays {
+		if delay.Duration <= 0 {
+			return fmt.Errorf("retry delay must be greater than zero, got %s. %w",
+				delay.Duration, ErrInvalidConfig)
+		}
+	}
+	if r.MaxAttempts < 0 {
+		return fmt.Errorf("max retries cannot be negative, got %d. %w",
+			r.MaxAttempts, ErrInvalidConfig)
```

Comment:
Does it make sense to first report the generic error and then fill in the details?

```suggestion
		return fmt.Errorf("%w: retry delays cannot be nil.", ErrInvalidConfig)
	}
	if len(r.Delays) == 0 {
		return fmt.Errorf("%w: retry delays cannot be empty.", ErrInvalidConfig)
	}
	for _, delay := range r.Delays {
		if delay.Duration <= 0 {
			return fmt.Errorf("%w: retry delay must be greater than zero, got %s.",
				delay.Duration)
		}
	}
	if r.MaxAttempts < 0 {
		return fmt.Errorf("%w: max retries cannot be negative, got %d",
			ErrInvalidConfig, r.MaxAttempts)
```

Reply:
No strong opinion about that, I apply it!